snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/MrcmTypeRequest.java

```diff
@@ -0,0 +1,160 @@
+
+/**
+ * @since 8.7.0
+ */
+public class MrcmTypeRequest implements Request<BranchContext, SnomedReferenceSetMembers> {
+	
+	private static final long serialVersionUID = 1L;
+
+	public enum ATTRIBUTE_TYPE {
+		DATA,
+		OBJECT,
+		ALL;
+		
+		public static ATTRIBUTE_TYPE getByNameIgnoreCase(String name) {
+			for (final ATTRIBUTE_TYPE type : values()) {
+				if (nullToEmpty(name).equalsIgnoreCase(type.toString())) {
+					return type;
+				}
+			}
+			throw new BadRequestException("Unknown attribute type: '%s'", name);
+		}
+	}
+	
+	private Set<String> selfIds;
+	private Set<String> ruleParentIds; 
+	private Set<String> refSetIds;
+	private List<String> moduleIds;
+	private ATTRIBUTE_TYPE attributeType;
+	
+	public void setSelfIds(Set<String> selfIds) {
+		this.selfIds = selfIds;
+	}
+
+	public void setParentIds(Set<String> ruleParentIds) {
+		this.ruleParentIds = ruleParentIds;
+	}
+
+	public void setRefSetIds(Set<String> refSetIds) {
+		this.refSetIds = refSetIds;
+	}
+
+	public void setModuleIds(List<String> moduleIds) {
+		this.moduleIds = moduleIds;
+	}
+
+	public void setAttributeType(ATTRIBUTE_TYPE attributeType) {
+		this.attributeType = attributeType;
+	}
+	
+	@Override
+	public SnomedReferenceSetMembers execute(BranchContext context) {
+		Set<String> inScopeRefSetIds = SnomedRequests.prepareSearchMember()
+				.all()
+				.filterByActive(true)
+				.filterByRefSetType(SnomedRefSetType.MRCM_MODULE_SCOPE)
+				.filterByReferencedComponent(moduleIds)
+				.setFields(ID, MRCM_RULE_REFSET_ID)
+				.build()
+				.execute(context)
+				.stream()
+				.map(m -> (String) m.getProperties().get(SnomedRf2Headers.FIELD_MRCM_RULE_REFSET_ID))
+				.collect(Collectors.toSet());
+		
+		if (inScopeRefSetIds.isEmpty()) {
+			return new SnomedReferenceSetMembers(0, 0);
+		}
+		
+		String eclConstraint = "*";
+		
+		switch (attributeType) {
+		case DATA: eclConstraint = String.format("<%s", CONCEPT_MODEL_DATA_ATTRIBUTE);
+			break;
+		case OBJECT: eclConstraint = String.format("<%s", CONCEPT_MODEL_OBJECT_ATTRIBUTE);
+			break;
+		case ALL: eclConstraint = String.format("<%s AND <%s", CONCEPT_MODEL_OBJECT_ATTRIBUTE, CONCEPT_MODEL_DATA_ATTRIBUTE );
+			break;		
+		};
+		
+		final Set<String> typeIds = SnomedRequests.prepareSearchConcept()
+			.all()
+			.filterByActive(true)
+			.filterByEcl(eclConstraint)
+			.setFields(SnomedConceptDocument.Fields.ID)
+			.build()
+			.execute(context)
+			.stream()
+			.map(SnomedConcept::getId)
+			.collect(Collectors.toSet());
+		
+		if (typeIds.isEmpty()) {
+			return new SnomedReferenceSetMembers(0, 0);
+		}
+		
+		Set<String> domainIds = new HashSet<>();
+		
+		if (selfIds != null) {
+			domainIds.addAll(selfIds);
+		}
+		
+		if (ruleParentIds != null) {
+			domainIds.addAll(ruleParentIds);
+		}
+		
+		if (refSetIds != null) {
+			domainIds.addAll(refSetIds);
+		}
+		
+		SnomedRefSetMemberSearchRequestBuilder requestBuilder = SnomedRequests.prepareSearchMember()
+			.all()
```

Comment:
Are we going to return all matching members all the time? What happens if we have hundreds of results? I would like to see a limit parameter introduced in the API to limit the number of types returned here and also a searchAfter parameter to control the paging.